lib/connectors_async/job_watcher.rb

```diff
@@ -0,0 +1,73 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+require 'concurrent'
+
+require 'connectors_shared/logger'
+
+module ConnectorsAsync
+  class JobWatcher
+    IDLE_TIME = 30 # seconds
+
+    class AlreadyWatchingError < StandardError; end
+
+    def initialize(job_store:)
+      @job_store = job_store
+
+      @pool = Concurrent::ThreadPoolExecutor.new(
```

Comment:
> Is a ThreadPool necessary here considering we only need one thread and there is only one task to execute?
No, I will change it to just use one thread, good call.

> When I first thought of how to clean stale jobs, I was thinking of cron job. May I know why you favor a watcher instead of a cron?

Having a cron job means that we will expose the API of jobs outside somehow, and it's probably not something we're willing to do now. Also having a cron means that connectors operations will be more complex - now you just do `make run` and forget, but with cron user will have to maintain the cron entry - install it, refresh it if something changed in cron definition, remove it manually, logs for this will be separated too, so a lot of work.

Also JobWatcher is a temporary abstraction - once we persist jobs we won't need it any more imo.

Reply:
Do you mean by cron having a separate record in `crontab` that executes some script to clean up jobs? Can you elaborate more about this?

> Could you elaborate on why it's not needed anymore when jobs are moved from memory to datastore? I think we don't need to clean up jobs when they are persisted. But we will need to clean up the documents queue, which should be separated from Job by then.

What I've meant is that the logic will change heavily, such as the code that cleans up stuff won't be needed much - in the new world we will still have some sort of the queue of course, but it will be detached from everything that's happening in current Job class. This queue will be working differently - there will be a worker that picks up data from it and tries to send to ElasticSearch as fast as possible, but the queue itself should never be cleaned up - it should be persisted in ElasticSearch, thus a case when queue is stuck - not sending documents - is an exception that should probably even terminate the process? This is something that we'll decide, but design of the ingestion process in the future will be just really really different in some parts, and I see the cleanup introduced here as something temporary, changing heavily in upcoming future.

Reply:
>Do you mean by cron having a separate record in `crontab` that executes some script to clean up jobs? 

We actually have a similar cron job in `ent-search`, Check [this](https://github.com/elastic/ent-search/blob/main/frito_pie/lib/work/cron/delete_old_connectors_jobs.rb)

>there will be a worker that picks up data from it and tries to send to ElasticSearch as fast as possible

Make sense. Forgot the new ingestion is push model

Reply:
> We actually have a similar cron job in ent-search, Check [this](https://github.com/elastic/ent-search/blob/main/frito_pie/lib/work/cron/delete_old_connectors_jobs.rb)
Oh wow, I totally forgot about these ones - thanks for sharing, it might also be something we could use in future when managing jobs!